README.md
# Paddle.com Node.js SDK

[![NPM version](https://img.shields.io/npm/v/paddle-sdk-dplnk.svg?style=flat-square)](https://www.npmjs.com/package/paddle-sdk-dplnk)

Welcome to the [Paddle.com](https://www.paddle.com/) Node.js SDK documentation.

## Installation

Install the SDK module using `npm`:

```
$ npm install paddle-sdk-dplnk
```

or using `yarn`:

```
$ yarn add paddle-sdk-dplnk
```

## Usage

Using `Promise`:

```js
const PaddleSDK = require('paddle-sdk-dplnk');

const client = new PaddleSDK('your-vendor-id-here', 'your-auth-code-here');

client.getProducts()
  .then(console.log)
  .then(() => client.getProductPlans(123))
  .then(console.log);
```

Using `async`/`await`:

```js
const PaddleSDK = require('paddle-sdk-dplnk');

async function run() {
  const client = new PaddleSDK('your-vendor-id-here', 'your-auth-code-here');

  const products = await client.getProducts();
  console.log(products);

	const plans = await client.getProductPlans(123);
  console.log(plans);
}

run();
```

## Documentation

See the JSDoc markdown documentation in [Documentation.md](Documentation.md).

## Change log

The change log can be found here: [CHANGELOG.md](CHANGELOG.md).

## ToDo

* Rewrite to typescript.
* Add missing endpoint

## Authors and license

Author: [Dmitry Dergunov](mailto:dmitry@dergunov.net).

Forked from: [Valentin Agachi](http://agachi.name/).

MIT License, see the included [License.md](LICENSE.md) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.2.1](https://github.com/dplnk/paddle-sdk/compare/v2.1.0...v2.2.1) (2020-04-13)


### Bug Fixes

* fix PaddleAPI params ([3b99c5a](https://github.com/dplnk/paddle-sdk/commit/3b99c5a42421a8ea9e5b2c742b52ebd9dfe0be1d))

## [2.2.0](https://github.com/dplnk/paddle-sdk/compare/v2.1.0...v2.2.0) (2020-04-13)

### Fixes
* Update params for Paddle-API

## [2.1.0](https://github.com/avaly/paddle-sdk/compare/v2.0.0...v2.1.0) (2019-12-29)


### Features

* Update subscription plan method ([310f54d](https://github.com/avaly/paddle-sdk/commit/310f54d932b3d4f6715c9ee668cdc6c43523ce4b))

## [2.0.0](https://github.com/avaly/paddle-sdk/compare/v1.3.0...v2.0.0) (2019-12-27)


### ⚠ BREAKING CHANGES

* Requires node v10+

### Features

* Upgrade all dependencies ([ce2d240](https://github.com/avaly/paddle-sdk/commit/ce2d2401681d1564e1c77f7904a98de0186fb85a))

## [1.3.0](https://github.com/avaly/paddle-sdk/compare/v1.2.2...v1.3.0) (2019-11-12)


### Features

* Add generatePayLink ([#8](https://github.com/avaly/paddle-sdk/issues/8)) ([a09b6c8](https://github.com/avaly/paddle-sdk/commit/a09b6c8))



### [1.2.2](https://github.com/avaly/paddle-sdk/compare/v1.2.1...v1.2.2) (2019-06-18)



<a name="1.2.1"></a>
## [1.2.1](https://github.com/avaly/paddle-sdk/compare/v1.2.0...v1.2.1) (2019-06-18)


### Bug Fixes

* Throw on error responses with 2xx status code ([28f07da](https://github.com/avaly/paddle-sdk/commit/28f07da))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/avaly/paddle-sdk/compare/v1.1.0...v1.2.0) (2018-03-23)


### Features

* Cancel subscription ([7a2a41c](https://github.com/avaly/paddle-sdk/commit/7a2a41c))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/avaly/paddle-sdk/compare/v1.0.0...v1.1.0) (2017-11-20)


### Features

* Verify Webhook Alerts :tada: ([d343da2](https://github.com/avaly/paddle-sdk/commit/d343da2))



<a name="1.0.0"></a>
# 1.0.0 (2017-11-19)


### Features

* Initial commit ([6bc14b0](https://github.com/avaly/paddle-sdk/commit/6bc14b0))
* Only getters methods are avilable in this version


sdk.js
const crypto = require('crypto');
const got = require('got');

const pkg = require('./package.json');
const serialize = require('./lib/serialize');

const SERVER_URL = 'https://vendors.paddle.com/api/2.0';

class PaddleSDK {
	/**
	 * @class PaddleSDK
	 * @typicalname client
	 * @param {string} vendorId - The vendor ID for a Paddle account
	 * @param {string} authCode - The Auth Code for a Paddle account
	 * @param {string} [publicKey] - The public key for a Paddle account used to verify webhooks, only required for `verifyWebhookData`
	 * @param {object} [options]
	 * @param {string} [options.server=vendors.paddle.com/api/2.0] - The server URL prefix for all requests
	 *
	 * @example
	 * const client = new PaddleSDK('your-vendor-id', 'your-auth-code');
	 * const client = new PaddleSDK('your-vendor-id', 'your-auth-code', 'your-public-key');
	 */
	constructor(vendorId, authCode, publicKey, options) {
		this.vendorId = vendorId || 'MISSING';
		this.authCode = authCode || 'MISSING';
		this.publicKey = publicKey || 'MISSING';
		this.server = (options && options.server) || SERVER_URL;
	}

	/**
	 * Execute a HTTP request
	 *
	 * @private
	 * @param path
	 * @param {string} url - url to do request
	 * @param {object} [options.body] - body parameters / object
	 * @param {object} [options.headers] - header parameters
	 * @param {boolean} [options.form] - form parameter (ref: got package)
	 * @param {boolean} [options.json] - json parameter (ref: got package)
	 */
	_request(path, { body = {}, headers = {}, form = true, json = false } = {}) {
		const url = this.server + path;
		const fullBody = Object.assign(body, this._getDefaultBody());

		const options = {
			headers: this._getDefaultHeaders(headers),
			method: 'POST',
		};
		if (form) {
			options.form = fullBody;
		}
		if (json) {
			options.json = fullBody;
		}

		return got(url, options)
			.json()
			.then(body => {
				if (typeof body.success === 'boolean') {
					if (body.success) {
						return body.response || body;
					}

					throw new Error(
						`Request ${url} returned an error! response=${JSON.stringify(body)}`
					);
				}

				return body;
			});
	}

	_getDefaultBody() {
		return {
			vendor_id: this.vendorId,
			vendor_auth_code: this.authCode,
		};
	}

	/**
	 * Get the list of required headers for an API request
	 *
	 * @private
	 * @param {object} [additionalHeaders={}] - headers object
	 */
	_getDefaultHeaders(additionalHeaders) {
		return Object.assign(
			{
				'User-Agent': `paddle-sdk-dplnk/${pkg.version} (${pkg.repository.url})`,
			},
			additionalHeaders || {}
		);
	}

	/**
	 * Get the current list of products
	 *
	 * @method
	 * @returns {Promise}
	 * @fulfil {object} - The products list
	 *
	 * @example
	 * const products = await client.getProducts();
	 */
	getProducts() {
		return this._request('/product/get_products');
	}

	/**
	 * Get the current list of coupons for a product
	 *
	 * @method
	 * @param {number} productId - The specific product/subscription ID.
	 * @returns {Promise}
	 * @fulfil {object} - The coupons list
	 *
	 * @example
	 * const coupons = await client.getProductCoupons(123);
	 */
	getProductCoupons(productId) {
		return this._request('/product/list_coupons', {
			body: { product_id: productId },
		});
	}

	/**
	 * Get the current list of plans for a subscription
	 *
	 * @method
	 * @param {number} [plan] - Filter: The product/plan ID
	 * @returns {Promise}
	 * @fulfil {object} - The plans list
	 *
	 * @example
	 * const plans = await client.getProductPlans(123);
	 */
	getProductPlans(plan) {
		return this._request('/subscription/plans', {
			body: { plan },
		});
	}

	/**
	 * Get the current list of users for a subscription plan
	 *
	 * @method
	 * @param [filter]
	 * @param {number} [filter.subscriptionId] - A specific user subscription ID
	 * @param {number} [filter.planId] - The subscription plan ID
	 * @param {'active' | 'past due' | 'trialling' | 'paused'} [filter.state] - The user subscription status.
	 * Returns all active, past due, trialling and paused subscription plans if not specified.
	 * @param [pagination]
	 * @param {number} [pagination.page] - Paginate return results
	 * @param {number} [pagination.resultPerPage] - Number of subscription records to return per page.
	 * @returns {Promise}
	 * @fulfil {object} - The users list
	 *
	 * @example
	 * const users = await client.getPlanUsers({subscriptionId: 123});
	 */
	getPlanUsers(filter, pagination) {
		return this._request('/subscription/users', {
			body: {
				subscriptionId: filter.subscriptionId,
				planId: filter.planId,
				state: filter.state,
				page: pagination.page,
				result_per_page: pagination.resultPerPage,
			},
		});
	}

	/**
	 * Get the list of payments for a subscription plan
	 *
	 * @method
	 * @param [filter]
	 * @param {number} [filter.subscriptionId]
	 * @param {number} [filter.plan]
	 * @param {0 | 1} [filter.isPaid] - Filter: Payment is paid (0 = No, 1 = Yes)
	 * @param {string} [filter.from] - Filter: Payments starting from (date in format YYYY-MM-DD)
	 * @param {string} [filter.to] - Filter: Payments up to (date in format YYYY-MM-DD)
	 * @param {boolean} [filter.isOneOffCharge]
	 * @returns {Promise}
	 * @fulfil {object} - The payments list
	 *
	 * @example
	 * const payments = await client.getPlanPayments(123);
	 */
	getPlanPayments(filter) {
		return this._request('/subscription/payments', {
			body: {
				subscription_id: filter.subscriptionId,
				plan: filter.plan,
				is_paid: filter.isPaid,
				from: filter.from,
				to: filter.to,
				is_one_off_charge: filter.isOneOffCharge,
			},
		});
	}

	/**
	 * Get the list of webhooks history
	 *
	 * @method
	 * @param [pagination]
	 * @param {number} [pagination.page] - Paginate returned results
	 * @param {string} [pagination.alertPerPage] - Number of webhook records to return per page
	 * @param {string} [pagination.queryHead] The date-time from which to begin the history
	 * @returns {Promise}
	 * @fulfil {object} - The webhooks history list
	 *
	 * @example
	 * const webhooksHistory = await client.getWebhooksHistory();
	 */
	getWebhooksHistory(pagination) {
		return this._request('/alert/webhooks', {
			body: {
				page: pagination.page,
				alerts_per_page: pagination.alertPerPage,
				query_head: pagination.queryHead,
			},
		});
	}

	/**
	 * Get the list of transations for a resource
	 *
	 * @private
	 * @param {string} entity - Filter: Entity type of the id
	 * @param {string} id - Filter: ID number for the specified entity
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 */
	_getTransactions(entity, id) {
		return this._request(`/${entity}/${id}/transactions`);
	}

	/**
	 * Get the list of transations for a user
	 *
	 * @method
	 * @param {number} userId
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const userTransactions = await client.getUserTransactions(123);
	 */
	getUserTransactions(userId) {
		return this._getTransactions('users', `${userId}`);
	}

	/**
	 * Get the list of transations for a subscription
	 *
	 * @method
	 * @param {number} subscriptionId
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const subscriptionTransactions = await client.getSubscriptionTransactions(123);
	 */
	getSubscriptionTransactions(subscriptionId) {
		return this._getTransactions('subscription', `${subscriptionId}`);
	}

	/**
	 * Get the list of transations for an order
	 *
	 * @method
	 * @param {number} orderId
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const orderTransactions = await client.getOrderTransactions(123);
	 */
	getOrderTransactions(orderId) {
		return this._getTransactions('order', `${orderId}`);
	}

	/**
	 * Get the list of transations for a checkout
	 *
	 * @method
	 * @param {number} checkoutID
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const checkoutTransactions = await client.getCheckoutTransactions(123);
	 */
	getCheckoutTransactions(checkoutID) {
		return this._getTransactions('checkout', `${checkoutID}`);
	}

	/**
	 * Verify a webhook alert data using signature and a public key to validate that
	 * it was indeed sent from Paddle.
	 *
	 * For more details: https://paddle.com/docs/reference-verifying-webhooks
	 *
	 * @method
	 * @param  {Object} postData The object with all the parameters sent to the webhook
	 * @return {boolean}
	 *
	 * @example
	 * const client = new PaddleSDK('your-vendor-id', 'your-auth-code', 'your-public-key');
	 *
	 * // inside an Express handler which uses express.bodyParser middleware
	 * const isVerified = client.verifyWebhookData(req.body);
	 */
	verifyWebhookData(postData) {
		const signature = postData.p_signature;

		const keys = Object.keys(postData)
			.filter(key => key !== 'p_signature')
			.sort();

		const sorted = {};
		keys.forEach(key => {
			sorted[key] = postData[key];
		});

		// PHP style serialize! :O
		const serialized = serialize(sorted);

		try {
			const verifier = crypto.createVerify('sha1');
			verifier.write(serialized);
			verifier.end();

			return verifier.verify(this.publicKey, signature, 'base64');
		} catch (err) {
			return false;
		}
	}

	// TODO: fix params
	/**
	 * Update (upgrade/downgrade) the plan of a subscription
	 *
	 * @method
	 * @param {number} subscriptionId
	 * @param {number} planId
	 * @param {boolean} prorate
	 * @returns {Promise}
	 * @fulfill {object} - The result of the operation
	 *
	 * @example
	 * const result = await client.updateSubscriptionPlan(123);
	 */
	updateSubscriptionPlan(subscriptionId, planId, prorate = false) {
		return this._request('/subscription/users/update', {
			body: {
				subscription_id: subscriptionId,
				plan_id: planId,
				prorate,
			},
		});
	}

	/**
	 * Cancels an active subscription
	 *
	 * @method
	 * @param {number} subscriptionId - The specific user subscription ID.
	 * @returns {Promise}
	 * @fulfil {object} - The result of the operation
	 *
	 * @example
	 * const result = await client.cancelSubscription(123);
	 */
	cancelSubscription(subscriptionId) {
		return this._request('/subscription/users_cancel', {
			body: { subscription_id: subscriptionId },
		});
	}

	/**
	 * Generate a custom pay link
	 *
	 * @method
	 * @param {object} body
	 * @returns {Promise}
	 * @fulfil {object} - The new pay link url
	 *
	 * @example
	 * const custom = await client.generatePayLink({
	 *  "title" : "my custom checkout",
	 *  "custom_message" : "some custom message"
	 * 	"prices": [
	 *		"USD:19.99",
	 *		"EUR:15.99"
	 *	 ]
	 *	});
	 */
	generatePayLink(body) {
		return this._request('/product/generate_pay_link', {
			body,
			form: false,
			json: true,
		});
	}
}

module.exports = PaddleSDK;


sdk.ts
// @ts-ignore
const crypto = require('crypto');
// TODO: use axios
const got = require('got');

const pkg = require('./package.json');
const serialize = require('./lib/serialize');

const SERVER_URL = 'https://vendors.paddle.com/api/2.0';

class PaddleSDK {
	private readonly vendorId;
	private readonly authCode;
	private readonly publicKey;
	private readonly server;

	/**
	 * @class PaddleSDK
	 * @typicalname client
	 * @param {string} vendorId - The vendor ID for a Paddle account
	 * @param {string} authCode - The Auth Code for a Paddle account
	 * @param {string} [publicKey] - The public key for a Paddle account used to verify webhooks, only required for `verifyWebhookData`
	 * @param {object} [options]
	 * @param {string} [options.server=vendors.paddle.com/api/2.0] - The server URL prefix for all requests
	 *
	 * @example
	 * const client = new PaddleSDK('your-vendor-id', 'your-auth-code');
	 * const client = new PaddleSDK('your-vendor-id', 'your-auth-code', 'your-public-key');
	 */
	constructor(vendorId, authCode, publicKey, options) {
		this.vendorId = vendorId || 'MISSING';
		this.authCode = authCode || 'MISSING';
		this.publicKey = publicKey || 'MISSING';
		this.server = (options && options.server) || SERVER_URL;
	}

	/**
	 * Execute a HTTP request
	 *
	 * @private
	 * @param path
	 * @param {string} url - url to do request
	 * @param {object} [options.body] - body parameters / object
	 * @param {object} [options.headers] - header parameters
	 * @param {boolean} [options.form] - form parameter (ref: got package)
	 * @param {boolean} [options.json] - json parameter (ref: got package)
	 */
	_request(path, { body = {}, headers = {}, form = true, json = false } = {}) {
		const url = this.server + path;
		const fullBody = Object.assign(body, this._getDefaultBody());

		const options = {
			headers: this._getDefaultHeaders(headers),
			method: 'POST',
		};
		if (form) {
			// @ts-ignore
			options.form = fullBody;
		}
		if (json) {
			// @ts-ignore
			options.json = fullBody;
		}

		return got(url, options)
			.json()
			.then(body => {
				if (typeof body.success === 'boolean') {
					if (body.success) {
						return body.response || body;
					}

					throw new Error(
						`Request ${url} returned an error! response=${JSON.stringify(body)}`
					);
				}

				return body;
			});
	}

	_getDefaultBody() {
		return {
			vendor_id: this.vendorId,
			vendor_auth_code: this.authCode,
		};
	}

	/**
	 * Get the list of required headers for an API request
	 *
	 * @private
	 * @param {object} [additionalHeaders={}] - headers object
	 */
	_getDefaultHeaders(additionalHeaders) {
		return Object.assign(
			{
				'User-Agent': `paddle-sdk-dplnk/${pkg.version} (${pkg.repository.url})`,
			},
			additionalHeaders || {}
		);
	}

	/**
	 * Get the current list of products
	 *
	 * @method
	 * @returns {Promise}
	 * @fulfil {object} - The products list
	 *
	 * @example
	 * const products = await client.getProducts();
	 */
	getProducts() {
		return this._request('/product/get_products');
	}

	/**
	 * Get the current list of coupons for a product
	 *
	 * @method
	 * @param {number} productId - The specific product/subscription ID.
	 * @returns {Promise}
	 * @fulfil {object} - The coupons list
	 *
	 * @example
	 * const coupons = await client.getProductCoupons(123);
	 */
	getProductCoupons(productId) {
		return this._request('/product/list_coupons', {
			body: { product_id: productId },
		});
	}

	/**
	 * Get the current list of plans for a subscription
	 *
	 * @method
	 * @param {number} [plan] - Filter: The product/plan ID
	 * @returns {Promise}
	 * @fulfil {object} - The plans list
	 *
	 * @example
	 * const plans = await client.getProductPlans(123);
	 */
	getSubscriptionPlans(plan) {
		return this._request('/subscription/plans', {
			body: { plan },
		});
	}

	/**
	 * Get the current list of users for a subscription plan
	 *
	 * @method
	 * @param [filter]
	 * @param {number} [filter.subscriptionId] - A specific user subscription ID
	 * @param {number} [filter.planId] - The subscription plan ID
	 * @param {'active' | 'past due' | 'trialling' | 'paused'} [filter.state] - The user subscription status.
	 * Returns all active, past due, trialling and paused subscription plans if not specified.
	 * @param [pagination]
	 * @param {number} [pagination.page] - Paginate return results
	 * @param {number} [pagination.resultPerPage] - Number of subscription records to return per page.
	 * @returns {Promise}
	 * @fulfil {object} - The users list
	 *
	 * @example
	 * const users = await client.getPlanUsers({subscriptionId: 123});
	 */
	getSubscriptionUsers(filter, pagination) {
		return this._request('/subscription/users', {
			body: {
				subscriptionId: filter?.subscriptionId,
				planId: filter?.planId,
				state: filter?.state,
				page: pagination?.page,
				result_per_page: pagination?.resultPerPage,
			},
		});
	}

	/**
	 * Get the list of payments for a subscription plan
	 *
	 * @method
	 * @param [filter]
	 * @param {number} [filter.subscriptionId]
	 * @param {number} [filter.plan]
	 * @param {0 | 1} [filter.isPaid] - Filter: Payment is paid (0 = No, 1 = Yes)
	 * @param {string} [filter.from] - Filter: Payments starting from (date in format YYYY-MM-DD)
	 * @param {string} [filter.to] - Filter: Payments up to (date in format YYYY-MM-DD)
	 * @param {boolean} [filter.isOneOffCharge]
	 * @returns {Promise}
	 * @fulfil {object} - The payments list
	 *
	 * @example
	 * const payments = await client.getPlanPayments(123);
	 */
	getSubscriptionPayments(filter) {
		return this._request('/subscription/payments', {
			body: {
				subscription_id: filter?.subscriptionId,
				plan: filter?.plan,
				is_paid: filter?.isPaid,
				from: filter?.from,
				to: filter?.to,
				is_one_off_charge: filter?.isOneOffCharge,
			},
		});
	}

	/**
	 * Get the list of webhooks history
	 *
	 * @method
	 * @param [pagination]
	 * @param {number} [pagination.page] - Paginate returned results
	 * @param {string} [pagination.alertPerPage] - Number of webhook records to return per page
	 * @param {string} [pagination.queryHead] The date-time from which to begin the history
	 * @returns {Promise}
	 * @fulfil {object} - The webhooks history list
	 *
	 * @example
	 * const webhooksHistory = await client.getWebhooksHistory();
	 */
	getWebhooksHistory(pagination) {
		return this._request('/alert/webhooks', {
			body: {
				page: pagination?.page,
				alerts_per_page: pagination?.alertPerPage,
				query_head: pagination?.queryHead,
			},
		});
	}

	/**
	 * Get the list of transations for a resource
	 *
	 * @private
	 * @param {string} entity - Filter: Entity type of the id
	 * @param {string} id - Filter: ID number for the specified entity
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 */
	_getTransactions(entity, id) {
		return this._request(`/${entity}/${id}/transactions`);
	}

	/**
	 * Get the list of transations for a user
	 *
	 * @method
	 * @param {number} userId
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const userTransactions = await client.getUserTransactions(123);
	 */
	getUserTransactions(userId) {
		return this._getTransactions('users', `${userId}`);
	}

	/**
	 * Get the list of transations for a subscription
	 *
	 * @method
	 * @param {number} subscriptionId
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const subscriptionTransactions = await client.getSubscriptionTransactions(123);
	 */
	getSubscriptionTransactions(subscriptionId) {
		return this._getTransactions('subscription', `${subscriptionId}`);
	}

	/**
	 * Get the list of transations for an order
	 *
	 * @method
	 * @param {number} orderId
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const orderTransactions = await client.getOrderTransactions(123);
	 */
	getOrderTransactions(orderId) {
		return this._getTransactions('order', `${orderId}`);
	}

	/**
	 * Get the list of transations for a checkout
	 *
	 * @method
	 * @param {number} checkoutID
	 * @returns {Promise}
	 * @fulfil {object} - The transations list
	 *
	 * @example
	 * const checkoutTransactions = await client.getCheckoutTransactions(123);
	 */
	getCheckoutTransactions(checkoutID) {
		return this._getTransactions('checkout', `${checkoutID}`);
	}

	/**
	 * Verify a webhook alert data using signature and a public key to validate that
	 * it was indeed sent from Paddle.
	 *
	 * For more details: https://paddle.com/docs/reference-verifying-webhooks
	 *
	 * @method
	 * @param  {Object} postData The object with all the parameters sent to the webhook
	 * @return {boolean}
	 *
	 * @example
	 * const client = new PaddleSDK('your-vendor-id', 'your-auth-code', 'your-public-key');
	 *
	 * // inside an Express handler which uses express.bodyParser middleware
	 * const isVerified = client.verifyWebhookData(req.body);
	 */
	verifyWebhookData(postData) {
		const signature = postData.p_signature;

		const keys = Object.keys(postData)
			.filter(key => key !== 'p_signature')
			.sort();

		const sorted = {};
		keys.forEach(key => {
			sorted[key] = postData[key];
		});

		// PHP style serialize! :O
		const serialized = serialize(sorted);

		try {
			// @ts-ignore
			const verifier = crypto.createVerify('sha1');
			verifier.write(serialized);
			verifier.end();

			return verifier.verify(this.publicKey, signature, 'base64');
		} catch (err) {
			return false;
		}
	}

	// TODO: fix params
	/**
	 * Update (upgrade/downgrade) the plan of a subscription
	 *
	 * @method
	 * @param {number} subscriptionId
	 * @param {number} planId
	 * @param {boolean} prorate
	 * @returns {Promise}
	 * @fulfill {object} - The result of the operation
	 *
	 * @example
	 * const result = await client.updateSubscriptionPlan(123);
	 */
	updateSubscriptionPlan(subscriptionId, planId, prorate = false) {
		return this._request('/subscription/users/update', {
			body: {
				subscription_id: subscriptionId,
				plan_id: planId,
				prorate,
			},
		});
	}

	/**
	 * Cancels an active subscription
	 *
	 * @method
	 * @param {number} subscriptionId - The specific user subscription ID.
	 * @returns {Promise}
	 * @fulfil {object} - The result of the operation
	 *
	 * @example
	 * const result = await client.cancelSubscription(123);
	 */
	cancelSubscription(subscriptionId) {
		return this._request('/subscription/users_cancel', {
			body: { subscription_id: subscriptionId },
		});
	}

	/**
	 * Generate a custom pay link
	 *
	 * @method
	 * @param {object} body
	 * @returns {Promise}
	 * @fulfil {object} - The new pay link url
	 *
	 * @example
	 * const custom = await client.generatePayLink({
	 *  "title" : "my custom checkout",
	 *  "custom_message" : "some custom message"
	 * 	"prices": [
	 *		"USD:19.99",
	 *		"EUR:15.99"
	 *	 ]
	 *	});
	 */
	generatePayLink(body) {
		return this._request('/product/generate_pay_link', {
			body,
			form: false,
			json: true,
		});
	}
}

module.exports = PaddleSDK;
